Custom tutorial notebooks

ChaamC opened the pull request:
New PR for the custom tutorial notebooks solution.
Adds jq/yq installation which is required to run the deploy-data-specific-image script from birdhouse-deploy repo.
Please refer to the [other PR](https://github.com/bird-house/birdhouse-deploy/pull/150) for more details

New solution that tries to solve what was tried to be done on this earlier [PR](https://github.com/bird-house/pavics-jupyter-base/pull/2).

cwcummings committed:
dockerfile now downloads tutorial notebooks when started

cwcummings committed:
added documentation

cwcummings committed:
rename variable in script

cwcummings committed:
revert first solution changes and add jq/yq installation to read yaml files from the deploy scripts

dbyrns reviewed (approved):
LGTM

cwcummings committed:
add deploy script and related env file

ChaamC commented:
Updated : deploy script and related env file added to the repo, and removed from birdhouse-deploy

tlvu commented on scheduler-jobs/deploy_data_pavics_jupyter.env:
```diff
@@ -0,0 +1,25 @@
+#############
+
+DEPLOY_DATA_JOB_SCHEDULE="${AUTODEPLOY_NOTEBOOK_FREQUENCY}"
+DEPLOY_DATA_JOB_JOB_NAME="notebookdeploy-${DEPLOY_DATA_JOB_JUPYTERHUB_IMAGE_NAME}"
```
I suggest, to make it simple, merge your `birdhouse/components/scheduler/deploy_data_specific_image_job.env ` with this one because you are rolling a custom cronjob generation code to match `deploy_data_specific_image` anyways.

I have a two-part `deploy_data_job.env` generic part and an specific override because it works for many jobs.  Yours work only for 1 specific job, might as well make it simple and have everything in this file, including the looping part so the activation in `env.local` is very simple for users.

tlvu reviewed (commented):
Really close here.  Only simplication is needed.  All the pieces seem to be here.

cwcummings committed:
merge env file with env found on birdhouse-deploy repo

cwcummings committed:
process dest_sub_dir parameter in script

ChaamC commented:
Updated PR : 
- moved the env file from birdhouse-deploy to this repo
- added a dest_sub_dir parameter to specify a subdirectory destination where to download the data

tlvu commented on scheduler-jobs/deploy_data_pavics_jupyter.env:
```diff
@@ -1,25 +1,90 @@
+
+    ##############################################################################
+    # End configuration vars
+    ##############################################################################
```
LOL, there's not start so might as well remove this "End configuration vars".

ChaamC replied:
Remains of the old copied code :P I will remove it!

dbyrns reviewed (approved):
Also perfect here!

tlvu commented on scheduler-jobs/deploy_data_specific_image:
```diff
@@ -9,6 +9,9 @@
 #   source_path: path of the desired file or folder in the source repo
+#   dest_sub_dir: Data is sent to a tutorial-notebooks directory with the same name as the image running the script.
+#            Use this parameter to specify a subdirectory path (without a prefix '/') in this directory where the data will be sent.
+#            If the parameter is absent or left empty, data is sent directly to the image's tutorial-notebooks folder
```
So before this `dest_sub_dir` fix, given `crim-eo` image, the notebooks would have been deployed at the root `/data/jupyterhub_user_data/tutorial-notebooks/` instead of `/data/jupyterhub_user_data/tutorial-notebooks/crim-eo`?

dbyrns replied:
No the host path mount inside the image is /data/jupyterhub_user_data/tutorial-notebooks/<image_name> so for your example /data/jupyterhub_user_data/tutorial-notebooks/crim-eo/
The sub_dir only allows to checkout multiple repository, for exemple : 
common notebooks with a dest_sub_dir "common" will be checkout to /data/jupyterhub_user_data/tutorial-notebooks/crim-eo/common
eo notebooks with a dest_sub_dir "eo" will be checkout to /data/jupyterhub_user_data/tutorial-notebooks/crim-eo/eo
If dest_sub_dir stay empty they will be checkout to /data/jupyterhub_user_data/tutorial-notebooks/crim-eo/

tlvu replied:
Ah I see, thanks.

tlvu commented on scheduler-jobs/deploy_data_pavics_jupyter.env:
```diff
@@ -0,0 +1,90 @@
+DEPLOY_DATA_JOB_CONFIG="/notebook_config.yml"
+
+# Default path where notebooks are deployed on the image running the script
+NOTEBOOK_DEST_DIR=/tmp/notebooks
```
Can you rename this to `DEPLOY_DATA_JOB_NOTEBOOK_DEST_DIR` to avoid potential future name clash.

tlvu replied:
Oh wait a sec, this is the default value but where is it setting the real value?

ChaamC replied:
I guess I could remove the "default" in the comment here, as the variable is only set here, and is not really made to be overwritten anywhere else.

tlvu replied:
So that value is by default `NOTEBOOK_DEST_DIR=/tmp/notebooks` but in reality we want it in `/data/jupyterhub_user_data/tutorial-notebooks/<image_name>`, so where is this real path being set?

ChaamC replied:
In this [code sample](https://github.com/bird-house/pavics-jupyter-base/blob/ebc86d5af1c919d78bc245effc269f5cea9686f8/scheduler-jobs/deploy_data_pavics_jupyter.env#L68), we mount the path such as `/data/jupyterhub_user_data/tutorial-notebooks/<image_name>` to the variable `DEPLOY_DATA_JOB_NOTEBOOK_DEST_DIR`

tlvu replied:
> In this [code sample](https://github.com/bird-house/pavics-jupyter-base/blob/ebc86d5af1c919d78bc245effc269f5cea9686f8/scheduler-jobs/deploy_data_pavics_jupyter.env#L68), we mount the path such as `/data/jupyterhub_user_data/tutorial-notebooks/<image_name>` to the variable `DEPLOY_DATA_JOB_NOTEBOOK_DEST_DIR`

OHH, yes you're right.

But just to make understanding the code  easier, I'd suggest setting that var to the proper value then use `--volume ${DEPLOY_DATA_JOB_NOTEBOOK_DEST_DIR}:${DEPLOY_DATA_JOB_NOTEBOOK_DEST_DIR}:rw`

ChaamC replied:
Code updated with your suggestion :)

tlvu reviewed (approved):
Looks good for me as well, except the part about the notebook dest dir.

Also, you've manually enabled this on a real PAVICS deployment and see it works properly?

ChaamC commented:
@tlvu 
> Also, you've manually enabled this on a real PAVICS deployment and see it works properly?

It worked on a local setup, but I will try to do a test with the iac too eventually.

matprov commented:
> ..but I will try to do a test with the iac too eventually.

@ChaamC Eventually like before merging this PR? :)

cwcummings committed:
minor fixes after PR feedback

cwcummings committed:
update vars in scripts

tlvu reviewed (approved):
All good for me with the last code update.

cwcummings committed:
fix pip warning problem on DockerHub

cwcummings committed:
rename destination folder to avoid conflict with other jobs

cwcummings committed:
add todo comment

cwcummings committed:
bump version date
